### ISSUE TITLE
Any reason not to create dynamic bindings package?

### ISSUE THREAD

**Issue by assumptionsoup** [AcademySoftwareFoundation/rez#537]
I think I understand why platform, os, and arch are bound packages: being real packages, they are discoverable. They use the same resolution architecture that the rest of rez uses (which means no surprises!). With the implicit soft requirements, I'm free to use them or create my own system if I choose to.

That being said, it seems a little strange to explicitly run rez-bind os/platform/os any time that the underlying system changes. Especially when rez packages may end up in the user profile, and migrate between computers. In particular, arch and os could change in this scenario, depending on how uniform your machines are at a given site. And explicit bindings will need to be regenerated every time the os is upgraded.

Aside from the additional time and computation costs, is there any reason I shouldn't make the platform, os, and arch packages dynamically resolve their version in the same way that rez implicitly requires them? It would make this particular system circular and contrived logic, but I don't seen any reason this wouldn't work. Do you?

Here's an example platform package.py implementing this idea:
```
# -*- coding: utf-8 -*-
name = 'platform'

def getVersion():
    from rez.system import system
    from rez.vendor.version.version import Version
    return str(Version(system.platform))

version = getVersion()

timestamp = 1540850024

format_version = 2
```

**Comment by nerdvegas** [comment]
See below.


On Tue, Oct 30, 2018 at 9:03 AM Jordan Hueckstaedt <notifications@github.com>
wrote:

> I think I understand why platform, os, and arch are bound packages: being
> real packages, they are discoverable. They use the same resolution
> architecture that the rest of rez uses (which means no surprises!). With
> the implicit soft requirements, I'm free to use them or create my own
> system if I choose to.
>
> That being said, it seems a little strange to explicitly run rez-bind
> os/platform/os any time that the underlying system changes. Especially when
> rez packages may end up in the user profile, and migrate between computers.
> In particular, arch and os might change in this scenario, depending on how
> uniform machines are at a given site. And explicit bindings will need to be
> regenerated every time the os is upgraded.
>

Well for the system you're consuming packages on, or even building on for
that matter, what matters is your implicit package list, which can be
dynamic (and is by default). You do need to create new platform/os/arch
packages if any new systems are getting built on, however this only needs
to be done once, regardless of how many devs/artists there are in a studio.
But I see your point if you're someone using rez at home for eg. Please
note though that rez-bind isn't really intended for production use, and I
want to deprecate it in favour of dynamic packages like the one you're
suggesting here (read on, however).

Your example platform package here will work in theory, but there is a
misunderstanding in its implementation. I assume this is an already
installed package (hence the timestamp being present)? I'm sure this is
just an example, but so you know, that package.py is not possible - package
attributes that are dynamic after installation have to be implemented as
"late binding" attributes, see here for details:
https://github.com/nerdvegas/rez/wiki/Package-Definition-Guide#package-attributes-as-functions

In any case, there is a problem implementing the version dynamically like
this. It does work, however it will cause cached resolves to not be aware
of a package's attributes having been changed. This then leads to
unpredictable behavior, where a resolve will use a package and yet be
unaware that that package's version has changed since the resolve occurred.
In this case you'd actually get away with it, assuming that
platform/os/arch are always present in the request (because they're
implicit requirements), But there's no hard guarantee of that either, since
implicit requirements are configurable. In any case, for these reasons,
late-bound 'version' is not supported.

The problem noted above applies to any attribute which can affect the
resolve in fact - "requires" is another example. In theory it's possible to
fix. Rez could do something like take a hash of all late bound attributes
in all packages in a cached resolve; then when a cached resolve is used,
re-evaluate all those attribs again, and check that everything still
matches. There are problems with this though. 1: the extra expense (now you
have to eval potentially a lot of attributes that you didn't touch at all
before); 2: it'll result in false negatives (ie unnecessary cache misses),
due to changed late-bound attribs that don't actually affect the resolve
(and it just isn't possible to tell what would or would not affect the
resolve - a late-bound 'requires' might refer to any attrib in the
package). 3: it'll result in a cache entry that is valid on one system, but
invalid on others. In fact the same cache entry would keep getting
overwritten due to resolves on different systems, causing a cache swapping
problem (kinda like how pyc files can get continually rewritten due to
differing bytecode python interpreters importing them). Again this could be
fixed (think: some double-hop type approach that gets the resolved cache
entry, then adds the late-bound hash value, then does another lookup based
on that). But all this gets quite complicated and still has significant
downsides. I don't know that there's an easy solution to this problem.

TL;DR - yes this could work in theory, but only based on a bunch of
assumptions that may be false for any given configured system. No easy
answer unfortunately.
